hibernate-core/src/main/java/org/hibernate/event/internal/DefaultAutoFlushEventListener.java

```diff
@@ -48,33 +52,41 @@ public void onAutoFlush(AutoFlushEvent event) throws HibernateException {
 				// Need to get the number of collection removals before flushing to executions
 				// (because flushing to executions can add collection removal actions to the action queue).
 				final ActionQueue actionQueue = source.getActionQueue();
-				final int oldSize = actionQueue.numberOfCollectionRemovals();
-				flushEverythingToExecutions( event );
-				if ( flushIsReallyNeeded( event, source ) ) {
-					LOG.trace( "Need to execute flush" );
-					event.setFlushRequired( true );
+				final AutoFlushEvent.FlushModeCheck flushModeCheck = event.getFlushModeCheck();
+				final EventSource session = event.getSession();
+				final PersistenceContext persistenceContext = session.getPersistenceContextInternal();
+				if ( flushModeCheck == ALL || flushModeCheck == PRE_FLUSH_ONLY ) {
+					preFlush( session, persistenceContext );
+				}
+				if ( flushModeCheck == ALL || flushModeCheck == NO_PRE_FLUSH ) {
+					final int oldSize = actionQueue.numberOfCollectionRemovals();
+					flushEverythingToExecutions( event, persistenceContext, session );
```

Comment:
Maybe I'm missing something, but isn't `flushEverythingToExecutions` the method that actually does the flush cascading which ultimately invokes the identifier generators? I.e. shouldn't this call be part of the `preFlush`?

Reply:
it seems this happens in the `prepareEntityFlushes`